kernel/log.h

```diff
@@ -69,6 +70,41 @@ YOSYS_NAMESPACE_BEGIN
 #define S__LINE__sub1(x) S__LINE__sub2(x)
 #define S__LINE__ S__LINE__sub1(__LINE__)
 
+// YS_DEBUGTRAP is a macro that is functionally equivalent to a breakpoint
+// if the platform provides such functionality, and does nothing otherwise.
+// If no debugger is attached, it starts a just-in-time debugger if available,
+// and crashes the process otherwise.
+#if defined(_WIN32)
+# define YS_DEBUGTRAP __debugbreak()
+#else
+# ifndef __has_builtin
+// __has_builtin is a GCC/Clang extension; on a different compiler (or old enough GCC/Clang)
+// that does not have it, using __has_builtin(...) is a syntax error.
+#  define __has_builtin(x) 0
+# endif
+# if __has_builtin(__builtin_debugtrap)
+#  define YS_DEBUGTRAP __builtin_debugtrap()
+# elif defined(__unix__)
+#  define YS_DEBUGTRAP raise(SIGTRAP)
+# else
+#  define YS_DEBUGTRAP do {} while(0)
+# endif
+#endif
+
+// YS_DEBUGTRAP_IF_DEBUGGING is a macro that is functionally equivalent to a breakpoint
+// if a debugger is attached, and does nothing otherwise.
+#if defined(_WIN32)
+# define YS_DEBUGTRAP_IF_DEBUGGING do { if (IsDebuggerPresent()) DebugBreak(); } while(0)
```

Comment:
Why `DebugBreak` here and `__debugbreak` above?  What's the difference between those?

Reply:
`__debugbreak` is an intrinsic, `DebugBreak` is a kernel32 function. `YS_DEBUGTRAP` tries to expand to a platform-independent intrinsic (so that it works on platforms like WASM or Graal well--I checked and both of them should prefer that).

Conversely, `YS_DEBUGTRAP_IF_DEBUGGING` is apriori platform dependent so it doesn't try that.

There's no actual difference in practice, I just wrote both of them idiomatically.

Reply:
Ah, alright.  Thanks for the explanation.